addons/resource_distribution_controller.go

```diff
@@ -0,0 +1,161 @@
+	r.Logger.Info("Setting up controller with manager")
+	labelSelectorPredicate, err := predicate.LabelSelectorPredicate(metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			utils.CreatedByLabelKey: "odf-multicluster-orchestrator",
+		},
+	})
+	if err != nil {
+		return err
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("resource_distribution_controller").
+		For(&corev1.ConfigMap{}, builder.WithPredicates(predicate.NewPredicateFuncs(func(object client.Object) bool {
+			return object.GetName() == utils.StorageClientMappingConfigMapName && object.GetNamespace() == r.CurrentNamespace
+		}))).
+		Watches(&templatev1.Template{}, handler.EnqueueRequestsFromMapFunc(func(_ context.Context, _ client.Object) []ctrl.Request {
+			return []ctrl.Request{reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: r.CurrentNamespace,
+					Name:      utils.StorageClientMappingConfigMapName,
+				},
+			}}
+		}), builder.WithPredicates(labelSelectorPredicate)).
+		WithEventFilter(predicate.NewPredicateFuncs(func(object client.Object) bool {
+			return object.GetNamespace() == r.CurrentNamespace
+		})).
+		Complete(r)
+}
+
+func (r *ResourceDistributionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Logger.With("Request", req)
+	logger.Info("Distributing resources to all StorageConsumers.")
+
+	var err error
+
+	var addonDeletionlock corev1.ConfigMap
+	err = r.SpokeClient.Get(ctx, types.NamespacedName{Namespace: r.CurrentNamespace, Name: AddonDeletionlockName}, &addonDeletionlock)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if !slices.Contains(addonDeletionlock.GetFinalizers(), ResourceDistributionFinalizer) {
+		logger.Info("Adding finalizer to addonDeletionlock", "finalizer", ResourceDistributionFinalizer)
+		addonDeletionlock.Finalizers = append(addonDeletionlock.Finalizers, ResourceDistributionFinalizer)
+		if err := r.SpokeClient.Update(ctx, &addonDeletionlock); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	addVRCToConsumers := []*ocsv1alpha1.StorageConsumer{}
+	storageConsumerList := ocsv1alpha1.StorageConsumerList{}
+	templateList := templatev1.TemplateList{}
+
+	err = r.SpokeClient.List(ctx, &storageConsumerList, client.InNamespace(r.CurrentNamespace))
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to distribute resources to StorageConsumers: %w", err)
+	}
+	if len(storageConsumerList.Items) == 0 {
+		logger.Info("No StorageConsumer to add/remove resources.")
+		if slices.Contains(addonDeletionlock.GetFinalizers(), ResourceDistributionFinalizer) {
+			logger.Info("Removing finalizer from addonDeletionlock", "finalizer", ResourceDistributionFinalizer)
+			addonDeletionlock.Finalizers = utils.RemoveString(addonDeletionlock.Finalizers, ResourceDistributionFinalizer)
+			if err := r.SpokeClient.Update(ctx, &addonDeletionlock); err != nil {
+				return ctrl.Result{}, err
+			}
```

Comment:
This logic is used several times across the code, better to wrap in a func in utils. Will make the flow clearer and avoid issues when modifying later